rasa_sdk/forms.py

```diff
@@ -726,57 +726,93 @@ async def extract_custom_slots(
             `SlotSet` for any extracted slots.
         """
         custom_slots = []
-        slots_to_extract = await self.required_slots(dispatcher, tracker, domain)
-
-        if slots_to_extract is None:
-            return []
+        slots_to_extract = await self.required_slots(
+            self.slots_mapped_in_domain(domain), dispatcher, tracker, domain
+        )
 
         for slot_name in slots_to_extract:
-            method_name = f"extract_{slot_name.replace('-', '_')}"
-            extract_method = getattr(self, method_name, None)
+            custom_slots += await self._extract_slot(
+                slot_name, dispatcher, tracker, domain
+            )
+        return custom_slots
 
-            if not extract_method:
-                warnings.warn(
-                    f"No method '{method_name}' found for slot "
-                    f"'{slot_name}'. Skipping extraction for this slot."
-                )
-                continue
+    async def _extract_slot(
+        self,
+        slot_name: Text,
+        dispatcher: "CollectingDispatcher",
+        tracker: "Tracker",
+        domain: "DomainDict",
+    ) -> List[EventType]:
+        method_name = f"extract_{slot_name.replace('-', '_')}"
+        also_mapped_in_domain = slot_name in self.slots_mapped_in_domain(domain)
+        extract_method = getattr(self, method_name, None)
+
+        if not extract_method and not also_mapped_in_domain:
+            warnings.warn(
+                f"No method '{method_name}' found for slot "
+                f"'{slot_name}'. Skipping extraction for this slot."
+            )
+            return []
+
+        if not extract_method:
```

Comment:
For these conditionals I find it hard to determine which combination of `also_mapped_in_domain` and `extract_method` lead to each branch as it requires keeping each previous `if` in your head. I think it would be clearer if you used either a traditional nested `if` structure, or re-stated both variables in each flat `if` statement.